examples/async/s3_cp.ml

```diff
@@ -0,0 +1,390 @@
+    let date_str = Date.to_string_iso8601_basic date' in
+    let date_key = mac (Cstruct.of_string ("AWS4"^secret_access_key)) date_str in
+    let date_region_key = mac date_key (string_of_region region) in
+    let date_region_service_key = mac date_region_key (string_of_service service) in
+    let signing_key = mac date_region_service_key "aws4_request" in
+    signing_key
+
+  let auth_request ?now ~hashed_payload ~region ~service ~aws_access_key ~aws_secret_key request =
+    (* Important use the same time for everything here *)
+    let time = match now with None -> (Time.now ()) | Some t -> t in
+    let date = Time.to_date ~zone:Time.Zone.utc time in
+    let (canonical_request, signed_headers) = canonical_request hashed_payload request in
+    let string_to_sign = string_to_sign ~time:time ~scope:(date, region) ~service canonical_request in
+    let signing_key = make_signing_key ~date ~region ~service ~secret_access_key:aws_secret_key in
+    let creds = Printf.sprintf "%s/%s/%s/%s/aws4_request"
+        aws_access_key (Date.to_string_iso8601_basic date)
+        (string_of_region region)
+        (string_of_service service)
+    in
+    let signature = Nocrypto.Hash.(mac `SHA256
+                                     ~key:signing_key
+                                     (Cstruct.of_string string_to_sign)) in
+    let auth_header = Printf.sprintf
+        "AWS4-HMAC-SHA256 Credential=%s,SignedHeaders=%s,Signature=%s"
+        creds signed_headers (Compat.cstruct_to_hex_string signature)
+    in
+    [("Authorization", auth_header);]
+
+end
+
+module S3 = struct
+
+  type conf = {
+    region : region;
+    aws_access_key : string;
+    aws_secret_key : string;
+  } with sexp
+
+  let make_request ?body conf ~meth ~bucket ~objekt =
+    (* )let open Cohttp.Request in*)
+    let host_str = region_host_string conf.region in
+    let uri = Printf.sprintf "https://%s/%s/%s" host_str bucket objekt
+              |> Uri.of_string in
+    let time = Time.now () in
+    (* If PUT add content length *)
+    let headers = match meth with
+      | `PUT -> begin
+          let length = Option.value_map ~f:(String.length) ~default:0 body in
+          [("Content-length", Int.to_string length)]
+        end
+      | _ -> [] in
+    let headers = headers @ [("Host", host_str)] in
+    let (amz_headers, hashed_payload) = Auth.make_amz_headers time ?body in
+    let headers = headers @ amz_headers in
+    let request = Request.make ~meth
+        ~headers:(Header.of_list headers)
+        uri in
+    let auth_header = Auth.auth_request ~now:time
+        ~hashed_payload ~region:conf.region ~service:`S3
+        ~aws_access_key:conf.aws_access_key
+        ~aws_secret_key:conf.aws_secret_key request
+    in
+    let headers' = headers @ auth_header in
+    Cohttp.Request.(request.headers <- (Header.of_list headers'));
+    match meth with
+    | `PUT -> Client.request
+                ~body:(Option.value ~default:"" body |> Body.of_string)
```

Comment:
`~default` should probably be ````Empty```. Although would be best if we'd expose that constructor as `Body.empty`